Proof of knowledge of representation

miha-stopar opened the pull request:
Signed-off-by: Miha <miha.stopar@xlab.si>

mancabizjak commented on crypto/representationproofs/representation.go:
```diff
@@ -0,0 +1,148 @@
+func ProveKnowledgeOfRepresentation() bool {
+	group, err := groups.NewSchnorrGroup(256)
+	if err != nil {
+		fmt.Println(err)
```
Please join L13 and L14 into a single `fmt.Printf()` call that prints both `err` and the message in the second call.

mancabizjak commented on crypto/representationproofs/representation.go:
```diff
@@ -0,0 +1,148 @@
+	secrets          []*big.Int
+	bases            []*big.Int
+	randomValues     []*big.Int
+	y 				 *big.Int
```
Fix indentation

mancabizjak commented on crypto/representationproofs/representation.go:
```diff
@@ -0,0 +1,148 @@
+		return nil, fmt.Errorf("Number of secrets and representation bases shoud be the same.")
+	}
+
+	var prover RepresentationProver
```
Replace L70-L78 with:

```go
return &RepresentationProver{
 Group:   group,
 secrets: secrets,
 bases:   bases,
 y:       y,
}, nil
````

mancabizjak commented on crypto/zkp/primitives/representationproofs/representation.go:
```diff
@@ -0,0 +1,165 @@
+	Group             *groups.ModCyclicGroup
+	bases            []*big.Int
+	proofRandomData  *big.Int
+	y 				 *big.Int
```
Fix indentation.

mancabizjak commented on crypto/zkp/primitives/representationproofs/representation.go:
```diff
@@ -0,0 +1,165 @@
+
+func NewRepresentationVerifier(group *groups.ModCyclicGroup, bases []*big.Int,
+		y *big.Int) *RepresentationVerifier {
+	verifier := RepresentationVerifier{
```
Fix indentation and replace L133-L138 with:
```go
return &RepresentationVerifier{
 Group:  group,
 base:   bases,
 y:      y,
} 
```

mancabizjak commented on crypto/zkp/primitives/representationproofs/representation.go:
```diff
@@ -0,0 +1,165 @@
+	secrets          []*big.Int
+	bases            []*big.Int
+	randomValues     []*big.Int
+	y 				 *big.Int
```
Fix indentation.

mancabizjak commented on crypto/zkp/primitives/representationproofs/representation.go:
```diff
@@ -0,0 +1,165 @@
+		return nil, fmt.Errorf("Number of secrets and representation bases shoud be the same.")
+	}
+
+	var prover RepresentationProver
```
Replace L87-L95 with:

```go
return &RepresentationProver{
 Group:    group,
 secrets:  secrets,
 bases:    bases,
 y:    	   y,
}, nil
```

mancabizjak commented on crypto/zkp/primitives/representationproofs/representation.go:
```diff
@@ -0,0 +1,165 @@
+/*
```
I just realised that `crypto/zkp/primitives/representationproofs/representation.go` and `crypto/representationproofs/representation.go` are exactly the same (except the first one has a license header); Is this is a mistake? If yes, please use just one and remove the other; as for the comments: please combine my review comments for both files (they might be a little scattered) and apply them to this one file.

mancabizjak commented on crypto/representationproofs/representation.go:
```diff
@@ -0,0 +1,148 @@
+		fmt.Printf("Error when generating SchnorrGroup")
+		return false
+	}
+	var bases []*big.Int
```
If we know in advance that a slice will have precisely 3 elements, why not just use a fixed-size array? 
Here's how L17-L22 would be if we used an array:

````go
var bases [3]*big.Int
for i := 0; i < len(bases); i++ {
 r := common.GetRandomInt(group.Q)
 base := group.Exp(group.G, r)
 bases[i] = base
}
````

The same goes for the `secrets` slice below.

mancabizjak commented on crypto/representationproofs/representation.go:
```diff
@@ -0,0 +1,148 @@
+	Group             *groups.ModCyclicGroup
+	bases            []*big.Int
+	proofRandomData  *big.Int
+	y 				 *big.Int
```
Fix indentation.

mancabizjak commented on crypto/representationproofs/representation.go:
```diff
@@ -0,0 +1,148 @@
+
+func NewRepresentationVerifier(group *groups.ModCyclicGroup, bases []*big.Int,
+		y *big.Int) *RepresentationVerifier {
+	verifier := RepresentationVerifier{
```
Please modify L116-L121 so that you just `return &RepresentationVerifier{...}`, and fix indentation.

mancabizjak commented on crypto/groups/modcyclicgroup.go:
```diff
@@ -0,0 +1,105 @@
+		return nil, err
+	}
+
+	return &ModCyclicGroup{
```
Please fix indentation (go-fmt !)

mancabizjak commented on crypto/groups/modcyclicgroup.go:
```diff
@@ -0,0 +1,105 @@
+// parameters P and Q where P = R * Q + 1 for some R. Order of G is Q.
+func NewSchnorrGroup(qBitLength int) (*ModCyclicGroup, error) {
+	group := ModCyclicGroup{}
+	g, q, p, err := common.GetSchnorrGroup(qBitLength)
```
Wouldn't it be better if you move the logic from `common.GetSchnorrGroup` directly to this `NewSchnorrGroup` function (and remove it from there)? I prefer to have all the logic related to schnorr groups inside one crypto package.

miha-stopar commented:
Thanks Manca, good observations! I will just briefly summarize here (I added comments to functions which should explain other questions):
* one representation.go was there left by mistake
* modcylicgroup.go renamed into schnorr.go
* yes, common/groups.go was meant to be deleted - I moved GetSchnorrGroup into groups/schorr.go and renamed it into GetSchnorrParams (the other content from common/groups.go will be removed in one of the next PRs)

miha-stopar commented:
Thanks Manca! All comments meaningful again. Hope now it is ok :)
